verification.py

```diff
@@ -445,14 +682,122 @@ def _run_seq_gen_msprime_comparison(self, model, length=100):
+                ),
+                "par": ["-m", "F84", "-t", "0.75", "-f", "0.25,0.25,0.25,0.25"],
+            },
+            "GTR": {
+                "model_id": msprime.GTR(
+                    relative_rates=[0.4, 0.1, 0.4, 0.2, 0.4, 0.4],
+                    root_distribution=[0.3, 0.2, 0.3, 0.2],
+                ),
+                "par": [
+                    "-m",
+                    "GTR",
+                    "-r",
+                    "0.4,0.1,0.4,0.2,0.4,0.4",
+                    "-f",
+                    "0.3,0.2,0.3,0.2",
+                ],
+            },
+        }
+
+        basedir = self._output_dir
+        num_replicates = 100
+        sg_results = collections.defaultdict(list)
+        ts_results = collections.defaultdict(list)
+        pos = [i for i in range(1, length + 1)]
+        mutation_rate = 1e-6
+        args = ["-q", "-s", str(mutation_rate), "-l", str(length), "-wa"]
+        args += model_dict[model]["par"]
+
+        alleles = ["A", "C", "G", "T"]
+        num_alleles = len(alleles)
+
+        for _ in tqdm.tqdm(range(num_replicates)):
+            ts = msprime.simulate(num_samples, Ne=1e4, length=length)
+            ts_mutated = msprime.mutate(
+                ts,
+                rate=mutation_rate,
+                model=model_dict[model]["model_id"],
+                discrete=True,
+            )
+            num_sites = ts_mutated.num_sites
+
+            triplets = collections.defaultdict(list)
+            for edge in ts.edges():
+                triplets[edge.parent].append(edge.child)
+
+            # run Seq-gen and calculate statistics
+            temp_result = self._run_seq_gen(
+                ts.first(),
+                args,
+                model_dict[model]["model_id"],
+                triplets,
+                alleles,
+                num_sites,
+            )
+            if temp_result:
+                c_sg, tm_chisq_sg, root_chisq_sg = temp_result
+                # mean pairwise diversity
+                pi_sg = allel.sequence_diversity(pos, c_sg)
+
+                sg_results["pi"].append(pi_sg)
+                sg_results["root_distribution"].append(root_chisq_sg)
+                sg_results["transition_matrix"].append(tm_chisq_sg)
+
+            # construct haplotypes for all nodes
+            sequences = self._sequences_internal_nodes(ts_mutated)
+
+            roots_d_ts = np.zeros((num_alleles,))
+            transitions_ts = np.zeros((num_alleles, num_alleles))
+            for site in sequences[0]:
+                roots_d_ts[alleles.index(site)] += 1
+            transitions_ts = self._transitions(sequences, triplets, alleles)
+            transition_matrix = model_dict[model]["model_id"].transition_matrix
+            tm_chisq_ts = self._transition_matrix_chi_sq(
+                transitions_ts, transition_matrix
+            )
+            exp_roots_d_ts = (
+                np.sum(roots_d_ts) * model_dict[model]["model_id"].root_distribution
```

Comment:
This is only the expectation if the chance of not mutating (the diagonal) is the same for all alleles, since sites with no mutations are not added.

Reply:
The actual probability that a partticular location has ancestral allele k is `mutation_rate * root_distribution[k] * (1 - exp(-transition_matrix[k,k] * total_branch_length))`. We could get the expected number of ancestral alleles by computing this iterating over the trees, like (untested):
```
expected_ancestral_states = np.zeros(num_alleles)
for t in ts.trees():
  expected_ancestral_states += mutation_rate * root_distribution * (1 - np.exp(-np.diag(transition_matrix)*t.total_branch_length))
```

Reply:
@petrelharp , a few more questions. I now realize I went head in first without thinking enough about whether I understood the concepts of the root distribution and the transition matrix enough (in terms of how they are being used by the machinery). Sorry about that. I was assuming the root distribution is the frequency distribution from which ancestral state is drawn as soon as a site gets hit by the first mutation. And that from then on, everything is determined by the transition matrix. 

So your code `exp_roots = num_muts * model.root_distribution` in the test_mutations.py modulee made total sense to me. This is what we expect to see in terms of ancestral states for the root of the tree. 

Now in the code you are proposing here, you calculate the probability of no mutation happening on a certain site for each of the samples, right? And I assume that `transition_matrix[k,k]` means `-row_sum(transition matrix)`, so the total probability of leaving a certain state? (Otherwise this is 0 for JukesCantor)

If root distribution is indeed what I think it is, doesn't it suffice to calculate the expectation as you were doing before in your test code?

Reply:
Whoops - I made a mistake above. Instead of the diagonal of the transition matrix I should have put the sum of the offdiagonal terms. Also, I had the mutation rate in the wrong place. I think it should be this:
```
expected_ancestral_states = np.zeros(num_alleles)
change_probs = transition_matrix.sum(axis=1) - np.diag(transition_matrix)
for t in ts.trees():
  expected_ancestral_states += root_distribution * (1 - np.exp(-mutation_rate * change_probs*t.total_branch_length))
```
... but, I see that I should draft up some documentation on how the mutation matrix models work. I'll do that now, which will hopefully answer the other questions.

Reply:
... and fixing this up to account for span:
```
expected_ancestral_states = np.zeros(num_alleles)
change_probs = transition_matrix.sum(axis=1) - np.diag(transition_matrix)

for t in ts.trees():
  if discrete:
    t_span = np.ceil(t.span[1]) - np.ceil(t.span[0])
  else:
    t_span = t.span[1] - t.span[0]
  expected_ancestral_states += root_distribution * t_span * (1 - np.exp(-mutation_rate * change_probs*t.total_branch_length))
```